### PR TITLE
Fixup some stuff with triu and tril for sparse matrices

### LINKED ISSUE #12649: Two sparse triu/tril in Base with different performance.
There are currently two `triu` and `tril` methods for sparse matrices. One in `csparse.jl` [here](https://github.com/JuliaLang/julia/blob/e637e75a38ade7a80fb1f867ac1b50fec1e2cb10/base/sparse/csparse.jl#L381) and one in `linalg.jl` [here](https://github.com/JuliaLang/julia/blob/e637e75a38ade7a80fb1f867ac1b50fec1e2cb10/base/sparse/linalg.jl#L267)

The latter one supports setting a diagonal offset `k` but for `k=0` the methods should be the same. Now timing these:

```julia
julia> A = sprand(10^5, 10^5, 25/10^5);

julia> @which triu(A)
triu(A::Base.SparseMatrix.SparseMatrixCSC{Tv,Ti<:Integer}) at sparse/csparse.jl:382

julia> @which triu(A, 0)
triu{Tv,Ti}(S::Base.SparseMatrix.SparseMatrixCSC{Tv,Ti}, k::Integer) at sparse/linalg.jl:268

julia> @time triu(A);
  0.294281 seconds (11 allocations: 38.957 MB, 37.89% gc time)

julia> @time triu(A, 0);
  0.067271 seconds (11 allocations: 19.843 MB, 12.05% gc time)

julia> triu(A) == triu(A,0)
true
```

means that the `triu` in `csparse.jl` (which is the default one and the one everyone is likely using) is slower (at least for this random matrix I tried, not sure how it varies with different sparse structure).

Doesn't it make sense to delete `triu` in `csparse.jl` and setting a default `k=0` on the one in `linalg.jl`? Same for `tril`.

### PR DESCRIPTION
**Pull request by KristofferC**
For some discussion leading to this PR, see #12649

This PR does a couple of things:
- Adds `@inbounds` and hoists the field loading in the `fkeep!` function (about 30% speedup)
- Adds a missing mutating version for `triu` and `tril` with a diagonal offset.
- Removes the non mutating versions that where using `fkeep!` since they are slower than the ones in `linalg.jl` due to having to copy the complete sparse matrix instead of just allocating what is needed.
- Not relevant to #12649 but it also adds non mutating methods to `droptol` and `dropzeros` to have parity with the surrounding functions.

Fixes #12649

### CODE REVIEW

File: base/sparse/linalg.jl [hunk 1]
```diff
@@ -264,7 +264,7 @@ end
 
 ## triu, tril
 
-function triu{Tv,Ti}(S::SparseMatrixCSC{Tv,Ti}, k::Integer)
+function triu{Tv,Ti}(S::SparseMatrixCSC{Tv,Ti}, k=0::Integer)
```
Comment (tkelman) [inline base/sparse/linalg.jl]: shouldn't these be `k::Integer = 0` ?

**Comment by KristofferC** [comment]
Added test now.

File: base/sparse/linalg.jl (hunk 1 above)
Comment (KristofferC) [inline base/sparse/linalg.jl]: yes of course, doh!

Comment (KristofferC) [inline base/sparse/linalg.jl]: updated

File: base/sparse/csparse.jl [hunk 2]
```diff
@@ -372,13 +375,27 @@ call(::DropTolFun, i,j,x,other) = abs(x)>other
...
 droptol!(A::SparseMatrixCSC, tol) = fkeep!(A, DropTolFun(), tol)
+droptol(A::SparseMatrixCSC) = droptol!(copy(A))
 dropzeros!(A::SparseMatrixCSC) = fkeep!(A, DropZerosFun(), nothing)
-triu!(A::SparseMatrixCSC) = fkeep!(A, TriuFun(), nothing)
-triu(A::SparseMatrixCSC) = triu!(copy(A))
-tril!(A::SparseMatrixCSC) = fkeep!(A, TrilFun(), nothing)
-tril(A::SparseMatrixCSC) = tril!(copy(A))
+dropzeros(A::SparseMatrixCSC) = dropzeros!(copy(A))
+
+function triu!(A::SparseMatrixCSC, k=0::Integer)
```
Comment (tkelman) [inline base/sparse/csparse.jl]: here too

Comment (KristofferC) [inline base/sparse/csparse.jl]: :+1:

**Comment by tkelman** [comment]
This looks good. Might be nice if the error messages were more descriptive, but I guess that doesn't absolutely have to be done right now.

**Comment by ViralBShah** [comment]
Do the existing tests have coverage for this?

**Comment by KristofferC** [comment]
Coverage for what? For `triu` and `tril` with a diagonal offset, yes. I added tests for `triu!` and `tril!` with diagonal offset.

**Comment by ViralBShah** [comment]
Thanks.

**Pull request merged**